In [ ]:
# Clase 28/07 sesión 3 módulo 6
# 1. Limpieza de Datos
import pandas as pd

# Dataset con valores nulos y duplicados
data = {
    'Nombre': ['Alice', 'Bob', 'Charlie', None, 'Alice'], # Se añade 'Alice' duplicada para el ejemplo
    'Edad': [25, None, 30, 22, 25],
    'Ciudad': ['Madrid', 'Barcelona', 'Madrid', 'Madrid', 'Madrid']
}

df = pd.DataFrame(data)
print("DataFrame Original:\n", df)

# Eliminar filas con valores nulos
# dropna() elimina cualquier fila que contenga al menos un valor nulo (None, NaN)
df_cleaned = df.dropna()
print("\nDataFrame después de eliminar nulos:\n", df_cleaned)

# Eliminar duplicados
# drop_duplicates() elimina filas idénticas
df_cleaned = df_cleaned.drop_duplicates()
print("\nDataFrame después de eliminar duplicados:\n", df_cleaned)

# Nota: En un caso real, la imputación de valores nulos (rellenar con media, mediana, moda)
# es otra técnica común antes de eliminar filas.

DataFrame Original:
     Nombre  Edad     Ciudad
0    Alice  25.0     Madrid
1      Bob   NaN  Barcelona
2  Charlie  30.0     Madrid
3     None  22.0     Madrid
4    Alice  25.0     Madrid

DataFrame después de eliminar nulos:
     Nombre  Edad  Ciudad
0    Alice  25.0  Madrid
2  Charlie  30.0  Madrid
4    Alice  25.0  Madrid

DataFrame después de eliminar duplicados:
     Nombre  Edad  Ciudad
0    Alice  25.0  Madrid
2  Charlie  30.0  Madrid


In [ ]:
# 2. Normalización (Min-Max Scaling)
# La normalización escala los datos a un rango específico (comúnmente entre 0 y 1). Esto es útil cuando las variables tienen diferentes
# escalas, lo que puede afectar el rendimiento de algunos modelos de Machine Learning
# 2. Normalización (Min-Max Scaling)
from sklearn.preprocessing import MinMaxScaler
import numpy as np

# Dataset con valores en diferentes escalas
# Por ejemplo, la primera columna podría ser "número de habitaciones" y la segunda "precio en miles de USD"
data = np.array([[10, 200],
                 [20, 300],
                 [30, 400]])
print("Datos Originales:\n", data)

# Aplicar Min-Max Scaling
#  Creamos un objeto MinMaxScaler para aplicar la normalización
scaler = MinMaxScaler()
# fit_transform() calcula los valores mínimos y máximos y luego escala los datos.
# Aplicamos la normalización con fit_transform():
# - fit(): encuentra el valor mínimo y máximo de cada columna
# - transform(): escala los datos a un rango de [0, 1]
data_normalized = scaler.fit_transform(data)
print("\nDatos Normalizados (Min-Max Scaling):\n", data_normalized)
# Escala los datos entre 0 y 1 para que todas las variables tengan la misma importancia (por ejemplo, en algoritmos
# como KNN (k-Nearest Neighbors) o regresión lineal/múltiple).
# Esto es útil porque algunos algoritmos son sensibles a la escala de los datos (si una variable tiene valores mucho
# más grandes que otra, puede dominar el modelo).

Datos Originales:
 [[ 10 200]
 [ 20 300]
 [ 30 400]]

Datos Normalizados (Min-Max Scaling):
 [[0.  0. ]
 [0.5 0.5]
 [1.  1. ]]


Explicación de Min-Max Scaling
Datos Originales

[[ 10 200]
 [ 20 300]
 [ 30 400]]
Este arreglo `numpy` representa un pequeño conjunto de datos con 3 observaciones (filas) y 2 variables (columnas). Por ejemplo:
Observación	Variable 1 (X1)	Variable 2 (X2)
1	10	200
2	20	300
3	30	400
Imaginemos que:
- X1 podría representar el número de habitaciones.
- X2 podría representar el precio en miles de dólares.
Datos Normalizados (Min-Max Scaling)

[[0.  0. ]
 [0.5 0.5]
 [1.  1. ]]
Este es el resultado luego de aplicar la técnica de escalamiento Min-Max, que transforma los valores de cada variable a un rango entre 0 y 1.
 ¿Cómo se calcula?
La fórmula del Min-Max Scaling es:

X_norm = (X - X_min) / (X_max - X_min)
Para la primera columna (X1: [10, 20, 30]):
- X_min = 10
- X_max = 30
Aplicando la fórmula:
- (10 - 10) / (30 - 10) = 0
- (20 - 10) / (30 - 10) = 10 / 20 = 0.5
- (30 - 10) / (30 - 10) = 20 / 20 = 1
Para la segunda columna (X2: [200, 300, 400]):
- X_min = 200
- X_max = 400
Aplicando la fórmula:
- (200 - 200) / (400 - 200) = 0
- (300 - 200) / (400 - 200) = 100 / 200 = 0.5
- (400 - 200) / (400 - 200) = 200 / 200 = 1
Resultado:
Observación	X1 Normalizado	X2 Normalizado
1	0.0	0.0
2	0.5	0.5
3	1.0	1.0
Esto garantiza que ambas variables están ahora en la misma escala, lo cual es importante cuando se usan algoritmos que dependen de la distancia (como KNN, SVM, etc.).


In [ ]:
# 3. Transformación de Datos (Logarítmica)
# Las transformaciones de datos se aplican para mejorar su distribución o reducir el impacto de valores extremos. La transformación
# logarítmica es útil para datos con una distribución sesgada.
# 3. Transformación de Datos (Logarítmica)
import numpy as np

# Datos con una distribución sesgada (ej. ingresos, poblaciones)
data = np.array([1, 10, 100, 1000, 10000]) # Añadimos un valor más grande para ilustrar
print("Datos Originales:\n", data)

# Aplicar una transformación logarítmica
# np.log() calcula el logaritmo natural (base e) de cada elemento
data_transformed = np.log(data)
print("\nDatos Transformados (Logarítmica):\n", data_transformed)

# Nota: Otras transformaciones incluyen la raíz cuadrada (np.sqrt) o Box-Cox.

Datos Originales:
 [    1    10   100  1000 10000]

Datos Transformados (Logarítmica):
 [0.         2.30258509 4.60517019 6.90775528 9.21034037]


In [ ]:
# 4. Codificación de Variables Categóricas
# Las variables categóricas (texto o etiquetas) deben convertirse a un formato numérico para que los modelos
# de Machine Learning puedan procesarlas.

In [ ]:
# a. Label Encoding
# Asigna un número único a cada categoría. Es útil cuando las categorías tienen un orden natural (jerarquía).
# 4a. Codificación de Variables Categóricas: Label Encoding
from sklearn.preprocessing import LabelEncoder
import pandas as pd # Se usa pandas para un ejemplo más claro con Series

# Ejemplo de datos categóricos (ej. niveles educativos)
datos = ['Pequeño', 'Mediano', 'Grande', 'Mediano', 'Pequeño']
print("Datos Originales (Categorías):\n", datos)

# Crear una instancia de LabelEncoder
le = LabelEncoder()

# Ajustar (fit) el encoder a los datos y transformarlos
# fit_transform() aprende las categorías únicas y las convierte en números
datos_transformados_le = le.fit_transform(datos)
print("\nDatos Transformados (Label Encoding):\n", datos_transformados_le)

# Para ver el mapeo de categorías a números:
print("\nMapeo de categorías a números:", dict(zip(le.classes_, le.transform(le.classes_))))

# Desventaja: Introduce una relación numérica artificial si las categorías no tienen un orden real.

Datos Originales (Categorías):
 ['Pequeño', 'Mediano', 'Grande', 'Mediano', 'Pequeño']

Datos Transformados (Label Encoding):
 [2 1 0 1 2]

Mapeo de categorías a números: {np.str_('Grande'): np.int64(0), np.str_('Mediano'): np.int64(1), np.str_('Pequeño'): np.int64(2)}


In [ ]:
# b. One-Hot Encoding
# Convierte cada categoría en una nueva columna binaria (0 o 1). Esto es ideal para categorías sin un orden natural, ya que evita
# introducir relaciones numéricas artificiales.
# 4b. Codificación de Variables Categóricas: One-Hot Encoding
from sklearn.preprocessing import OneHotEncoder
import numpy as np

# Ejemplo de datos categóricos (ej. colores, no tienen un orden)
datos = np.array([["Rojo"], ["Verde"], ["Azul"], ["Rojo"]])
print("Datos Originales (Categorías):\n", datos)

# Crear una instancia de OneHotEncoder
# handle_unknown='ignore' previene errores si aparecen categorías nuevas en el futuro
# sparse_output=False (o .toarray()) asegura una salida como array denso
ohe = OneHotEncoder(handle_unknown='ignore', sparse_output=False)
# andle_unknown='ignore':
# Este es un parámetro muy útil del OneHotEncoder. Significa cómo debe comportarse el codificador si, después de haber sido
# entrenado con un conjunto de datos, se encuentra con una categoría que no vio durante el entrenamiento al transformar nuevos datos.
# Si lo dejas en su valor predeterminado ('error'), el codificador lanzaría un error si aparece una categoría desconocida.
# Al establecerlo en 'ignore', el codificador simplemente ignorará esa categoría desconocida y sus columnas correspondientes serán todas
# ceros. Esto es útil en producción para evitar que tu modelo se caiga por datos inesperados, aunque es importante estar al tanto de que
# esas categorías no estarán siendo representadas.

# Cuando se aplica One-Hot Encoding, si tienes muchas categorías, se crean muchas columnas nuevas, la mayoría de las cuales contendrán ceros.
# Esto puede resultar en matrices muy grandes y con muchos ceros, conocidas como matrices dispersas (sparse matrices).

# Por defecto, OneHotEncoder solía devolver una matriz dispersa por razones de eficiencia de memoria, especialmente cuando se trabaja con datasets
# muy grandes con muchas categorías.

# Al establecer sparse_output=False, le indicas al codificador que convierta la salida a un array de NumPy denso (completo), donde todos los ceros
# se almacenan explícitamente. Esto es más fácil de visualizar y trabajar para la mayoría de los casos de uso, especialmente en ejemplos didácticos
# o con menos categorías. (La alternativa sería usar .toarray() después de la transformación si sparse_output fuera True).
# En resumen, esta línea está configurando el codificador para que sea robusto ante nuevas categorías y para que la salida sea un array de NumPy fácil de usar.

# Ajustar (fit) el encoder a los datos y transformarlos
datos_transformados_ohe = ohe.fit_transform(datos)
print("\nDatos Transformados (One-Hot Encoding):\n", datos_transformados_ohe)

# Para ver los nombres de las nuevas columnas (características)
print("\nNombres de las nuevas columnas:", ohe.get_feature_names_out(['Color']))

# Desventaja: Aumenta la dimensionalidad si hay muchas categorías únicas.

Datos Originales (Categorías):
 [['Rojo']
 ['Verde']
 ['Azul']
 ['Rojo']]

Datos Transformados (One-Hot Encoding):
 [[0. 1. 0.]
 [0. 0. 1.]
 [1. 0. 0.]
 [0. 1. 0.]]

Nombres de las nuevas columnas: ['Color_Azul' 'Color_Rojo' 'Color_Verde']


Explicación de One-Hot Encoding
Datos Originales (Categorías):
[['Rojo'], ['Verde'], ['Azul'], ['Rojo']]
Este es un arreglo de categorías, en este caso colores. Tenemos cuatro observaciones, donde cada observación tiene un solo color como valor.
Objetivo del One-Hot Encoding:
Transformar variables categóricas (como 'Rojo', 'Verde', 'Azul') en variables numéricas binarias, que los algoritmos de Machine Learning puedan utilizar.
Resultado del One-Hot Encoding:
[[0. 1. 0.]  → Rojo
 [0. 0. 1.]  → Verde
 [1. 0. 0.]  → Azul
 [0. 1. 0.]] → Rojo
Cada color se convierte en una columna binaria:
- Si la observación tiene ese color, el valor es 1.
- Si no lo tiene, el valor es 0.
 Nombres de las nuevas columnas:
['Color_Azul', 'Color_Rojo', 'Color_Verde']
Estas columnas representan cada una de las categorías originales:
- 'Color_Azul' es 1 si el color es azul.
- 'Color_Rojo' es 1 si el color es rojo.
- 'Color_Verde' es 1 si el color es verde.
Ejemplo con interpretación:
Color original	Color_Azul	Color_Rojo	Color_Verde
Rojo	0	1	0
Verde	0	0	1
Azul	1	0	0
Rojo	0	1	0


In [ ]:
# c. Variables Dummy (con Pandas)
# Similar a One-Hot Encoding, pero se implementa directamente en DataFrames de Pandas, lo que la hace muy conveniente
# para datos tabulares.
# 4c. Codificación de Variables Categóricas: Variables Dummy (con Pandas)
import pandas as pd

# Crear un DataFrame de ejemplo con una columna categórica
data = {'Producto': ['Laptop', 'Mouse', 'Teclado', 'Laptop', 'Mouse'],
        'Precio': [1200, 25, 75, 1300, 30]}
df_productos = pd.DataFrame(data)
print("DataFrame Original:\n", df_productos)

# Aplicar get_dummies para crear variables dummy
# drop_first=True es común para evitar la "trampa de la variable dummy" (redundancia)
df_dummies = pd.get_dummies(df_productos, columns=['Producto'], drop_first=True, dtype=int)
print("\nDataFrame con Variables Dummy:\n", df_dummies)

# Observa que se crearon nuevas columnas para cada categoría del 'Producto'.
# Si drop_first=False, se crearían N columnas para N categorías.
# Con drop_first=True, se crean N-1 columnas, ya que la última categoría puede deducirse.

DataFrame Original:
   Producto  Precio
0   Laptop    1200
1    Mouse      25
2  Teclado      75
3   Laptop    1300
4    Mouse      30

DataFrame con Variables Dummy:
    Precio  Producto_Mouse  Producto_Teclado
0    1200               0                 0
1      25               1                 0
2      75               0                 1
3    1300               0                 0
4      30               1                 0


pd.get_dummies(...):

Esta es la función de la librería pandas (abreviada como pd) que se encarga de crear las variables dummy. Es una forma muy conveniente y directa de realizar One-Hot Encoding en un DataFrame.

df_productos:

Este es el primer argumento y se refiere al DataFrame de Pandas original que contiene la(s) columna(s) categórica(s) que deseas transformar. En tu ejemplo, es el DataFrame df_productos que creaste.

columns=['Producto']:

Este argumento especifica cuáles columnas del DataFrame deben ser convertidas en variables dummy. Debes pasar una lista de nombres de columnas. En este caso, solo queremos transformar la columna llamada 'Producto'. Si tuvieras más columnas categóricas que quisieras transformar, las incluirías en esta lista (ej. columns=['Producto', 'Color', 'Ciudad']).

drop_first=True:

Este es un parámetro muy importante. Cuando lo estableces en True, pandas elimina la primera categoría de cada variable al crear las columnas dummy.

¿Por qué es importante? Esto se hace para evitar la "trampa de la variable dummy" o multicolinealidad perfecta. Si tuvieras N categorías para una variable, get_dummies normalmente crearía N nuevas columnas binarias. Sin embargo, la información de una de esas columnas puede ser deducida a partir de las otras N-1 columnas. Por ejemplo, si tienes "Rojo", "Verde", "Azul" y creas tres columnas, si "Rojo" es 0 y "Verde" es 0, automáticamente sabes que es "Azul". Eliminar una de ellas (la primera por convención) evita esta redundancia y ayuda a algunos modelos (como la regresión lineal) a funcionar correctamente.

dtype=int:

Este parámetro especifica el tipo de dato de las nuevas columnas dummy. Al establecerlo en int, las nuevas columnas binarias contendrán 0 y 1 como enteros, lo cual es lo más común y eficiente para estas variables.

df_dummies = ...:

Finalmente, el resultado de la transformación (un nuevo DataFrame con las variables dummy añadidas y las columnas originales categóricas eliminadas) se asigna a una nueva variable, df_dummies.

In [ ]:
# 5. Escalamiento con StandardScaler
# El StandardScaler transforma los datos restando la media y dividiendo por la desviación estándar. Esto resulta en una
# distribución con media 0 y desviación estándar 1, lo cual es útil para algoritmos que asumen una distribución normal de
# los datos o para aquellos que son sensibles a la escala de las características (como redes neuronales y algoritmos basados en gradiente).
# 5. Escalamiento con StandardScaler
from sklearn.preprocessing import StandardScaler
import numpy as np

# Dataset con valores en diferentes escalas (igual que el ejemplo de Min-Max)
data = np.array([[10, 200],
                 [20, 300],
                 [30, 400],
                 [40, 500]]) # Añadimos un cuarto punto para tener más datos

print("Datos Originales:\n", data)

# Aplicar StandardScaler
scaler = StandardScaler()
# fit_transform() calcula la media y la desviación estándar y luego escala los datos
data_standardized = scaler.fit_transform(data)
print("\nDatos Estandarizados (StandardScaler):\n", data_standardized)

# Explicación de la salida:
# La media de cada columna se ha centrado en 0 y su desviación estándar es 1.
# Observa que los valores ahora pueden ser negativos.

Datos Originales:
 [[ 10 200]
 [ 20 300]
 [ 30 400]
 [ 40 500]]

Datos Estandarizados (StandardScaler):
 [[-1.34164079 -1.34164079]
 [-0.4472136  -0.4472136 ]
 [ 0.4472136   0.4472136 ]
 [ 1.34164079  1.34164079]]


Explicación de Estandarización con StandardScaler
Datos Originales:
[[ 10 200]
 [ 20 300]
 [ 30 400]
 [ 40 500]]
Descripción de las columnas:
Supongamos que estas dos columnas representan variables distintas, como por ejemplo:
- La primera columna: cantidad de productos vendidos.
- La segunda columna: ingresos generados por ventas.
¿Qué hace StandardScaler?
StandardScaler transforma los datos para que tengan media = 0 y desviación estándar = 1.
Se aplica la fórmula: Z = (X - μ) / σ
donde μ es la media y σ es la desviación estándar de la columna.
Cálculo para la primera columna:
Valores: 10, 20, 30, 40
Media (μ) = (10 + 20 + 30 + 40) / 4 = 25
Desviación estándar (σ) = sqrt(((10-25)^2 + (20-25)^2 + (30-25)^2 + (40-25)^2)/3) ≈ 16.43

Transformación de cada valor:
- (10 - 25) / 16.43 ≈ -1.34
- (20 - 25) / 16.43 ≈ -0.45
- (30 - 25) / 16.43 ≈ 0.45
- (40 - 25) / 16.43 ≈ 1.34
Datos Estandarizados (StandardScaler):
[[-1.34164079 -1.34164079]
 [-0.4472136  -0.4472136 ]
 [ 0.4472136   0.4472136 ]
 [ 1.34164079  1.34164079]]
Interpretación:
- Cada valor indica cuántas desviaciones estándar se encuentra respecto al promedio.
- Valores negativos indican que están por debajo del promedio.
- Valores positivos indican que están por encima del promedio.
- Como ambas columnas tienen la misma estructura (misma escala), sus valores estandarizados son idénticos.


In [ ]:
# Segunda parte
# Paso 1: Importar librerías
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, StandardScaler, MinMaxScaler
from sklearn.compose import ColumnTransformer

# Paso 2: Crear datos simulados
data = {
    'sexo': ['M', 'F', 'F', 'M', 'F'],
    'nivel': ['alto', 'medio', 'bajo', 'medio', 'alto'],
    'edad': [25, 32, 47, 51, 38],
    'ingreso': [3000, 4000, 2000, 5000, 3500]
}

df = pd.DataFrame(data)
print(" Datos originales:")
print(df)

# Paso 3: Codificación con LabelEncoder (para una sola columna)
le = LabelEncoder()
df['sexo_codificado'] = le.fit_transform(df['sexo'])

print("\n Codificación LabelEncoder para 'sexo':")
print(df[['sexo', 'sexo_codificado']])

 Datos originales:
  sexo  nivel  edad  ingreso
0    M   alto    25     3000
1    F  medio    32     4000
2    F   bajo    47     2000
3    M  medio    51     5000
4    F   alto    38     3500

🔤 Codificación LabelEncoder para 'sexo':
  sexo  sexo_codificado
0    M                1
1    F                0
2    F                0
3    M                1
4    F                0


In [ ]:
# Paso 4: Codificación con OneHotEncoder
ohe = ColumnTransformer(
    transformers=[
        ('sexo_ohe', OneHotEncoder(drop='first'), ['sexo']),
        ('nivel_ohe', OneHotEncoder(drop='first'), ['nivel'])
    ],
    remainder='passthrough'
)

df_ohe = ohe.fit_transform(df)
df_ohe = pd.DataFrame(df_ohe, columns=ohe.get_feature_names_out())
print("\n Codificación OneHotEncoder para 'sexo' y 'nivel':")
print(df_ohe)


 Codificación OneHotEncoder para 'sexo' y 'nivel':
   sexo_ohe__sexo_M  nivel_ohe__nivel_bajo  nivel_ohe__nivel_medio  \
0               1.0                    0.0                     0.0   
1               0.0                    0.0                     1.0   
2               0.0                    1.0                     0.0   
3               1.0                    0.0                     1.0   
4               0.0                    0.0                     0.0   

   remainder__edad  remainder__ingreso  remainder__sexo_codificado  \
0             25.0              3000.0                         1.0   
1             32.0              4000.0                         0.0   
2             47.0              2000.0                         0.0   
3             51.0              5000.0                         1.0   
4             38.0              3500.0                         0.0   

   remainder__edad_std  remainder__ingreso_minmax  
0            -1.428496                   0.333333  
1 

In [ ]:
# Paso 5: Escalamiento de variables numéricas
scaler_std = StandardScaler()
scaler_minmax = MinMaxScaler()

df['edad_std'] = scaler_std.fit_transform(df[['edad']])
df['ingreso_minmax'] = scaler_minmax.fit_transform(df[['ingreso']])

print("\n Escalamiento de edad con StandardScaler y de ingreso con MinMaxScaler:")
print(df[['edad', 'edad_std', 'ingreso', 'ingreso_minmax']])


 Escalamiento de edad con StandardScaler y de ingreso con MinMaxScaler:
   edad  edad_std  ingreso  ingreso_minmax
0    25 -1.428496     3000        0.333333
1    32 -0.693241     4000        0.666667
2    47  0.882306     2000        0.000000
3    51  1.302452     5000        1.000000
4    38 -0.063022     3500        0.500000


In [ ]:
# Tercer ejemplo
#  Importar librerías
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, MinMaxScaler, StandardScaler
from sklearn.metrics import pairwise_distances

#  Crear un DataFrame con datos de ejemplo
data = {
    'Nombre': ['Ana', 'Luis', 'Carlos', 'Marta', 'Pedro'],
    'Nivel_Educativo': ['Bajo', 'Medio', 'Alto', 'Medio', 'Bajo'],
    'Edad': [22, 35, 45, 28, 50],
    'Ingreso': [1500, 2500, 3200, 2100, 2700]
}
df = pd.DataFrame(data)
print("🔹 Datos Originales:")
print(df)

#  Codificación Label Encoding (para variables ordinales)
label_encoder = LabelEncoder()
df['Nivel_Label'] = label_encoder.fit_transform(df['Nivel_Educativo'])
print("\n🔹 Label Encoding:")
print(df[['Nivel_Educativo', 'Nivel_Label']])


🔹 Datos Originales:
   Nombre Nivel_Educativo  Edad  Ingreso
0     Ana            Bajo    22     1500
1    Luis           Medio    35     2500
2  Carlos            Alto    45     3200
3   Marta           Medio    28     2100
4   Pedro            Bajo    50     2700

🔹 Label Encoding:
  Nivel_Educativo  Nivel_Label
0            Bajo            1
1           Medio            2
2            Alto            0
3           Medio            2
4            Bajo            1


In [ ]:
# Codificación One-Hot Encoding
onehot_encoder = OneHotEncoder(sparse_output=False)
nivel_encoded = onehot_encoder.fit_transform(df[['Nivel_Educativo']])
columns_ohe = onehot_encoder.get_feature_names_out(['Nivel_Educativo'])
df_ohe = pd.DataFrame(nivel_encoded, columns=columns_ohe)
print("\n🔹 One-Hot Encoding:")
print(df_ohe)



🔹 One-Hot Encoding:
   Nivel_Educativo_Alto  Nivel_Educativo_Bajo  Nivel_Educativo_Medio
0                   0.0                   1.0                    0.0
1                   0.0                   0.0                    1.0
2                   1.0                   0.0                    0.0
3                   0.0                   0.0                    1.0
4                   0.0                   1.0                    0.0


In [ ]:
#  Variables Dummy con pandas
df_dummy = pd.get_dummies(df['Nivel_Educativo'], prefix='Nivel')
print("\n🔹 Variables Dummy (pandas):")
print(df_dummy)


🔹 Variables Dummy (pandas):
   Nivel_Alto  Nivel_Bajo  Nivel_Medio
0       False        True        False
1       False       False         True
2        True       False        False
3       False       False         True
4       False        True        False


In [ ]:
#  Unir todas las transformaciones al dataframe original
df_total = pd.concat([df, df_ohe, df_dummy], axis=1)
print("\n DataFrame con todas las codificaciones:")
print(df_total)


 DataFrame con todas las codificaciones:
   Nombre Nivel_Educativo  Edad  Ingreso  Nivel_Label  Edad_Scale  \
0     Ana            Bajo    22     1500            1    0.000000   
1    Luis           Medio    35     2500            2    0.464286   
2  Carlos            Alto    45     3200            0    0.821429   
3   Marta           Medio    28     2100            2    0.214286   
4   Pedro            Bajo    50     2700            1    1.000000   

   Ingreso_Scale  Edad_Std  Ingreso_Std  Nivel_Educativo_Alto  \
0       0.000000 -1.349652    -1.571468                   0.0   
1       0.588235 -0.096404     0.174608                   0.0   
2       1.000000  0.867634     1.396861                   1.0   
3       0.352941 -0.771230    -0.523823                   0.0   
4       0.705882  1.349652     0.523823                   0.0   

   Nivel_Educativo_Bajo  Nivel_Educativo_Medio  Nivel_Alto  Nivel_Bajo  \
0                   1.0                    0.0       False        True   
1   

Explicación del DataFrame con todas las codificaciones
Columnas originales
• Nombre: Nombre del individuo (dato categórico, no transformado).
• Nivel_Educativo: Nivel educativo original: Bajo, Medio o Alto.
• Edad: Edad numérica original.
• Ingreso: Ingreso numérico original.

Codificaciones y transformaciones
• Nivel_Label: Codificación de etiquetas (Label Encoding). Asigna un número entero a cada categoría de Nivel_Educativo:
  – Alto → 0
  – Bajo → 1
  – Medio → 2
• Edad_Scale: Edad normalizada (Min-Max Scaling). Escala Edad a un rango entre 0 y 1 usando: (x - min) / (max - min)
• Ingreso_Scale: Ingreso normalizado (Min-Max Scaling).
• Edad_Std: Edad estandarizada (Z-score): (x - media) / desviación estándar.
• Ingreso_Std: Ingreso estandarizado (Z-score).

One-Hot Encoding (variables binarias)
• Nivel_Educativo_Alto: 1 si el nivel educativo es Alto, 0 si no.
• Nivel_Educativo_Bajo: 1 si el nivel educativo es Bajo, 0 si no.
• Nivel_Educativo_Medio: 1 si el nivel educativo es Medio, 0 si no.

Codificación booleana (True/False)
• Nivel_Alto: True si el nivel educativo es Alto, False si no.
• Nivel_Bajo: True si es Bajo, False si no.
• Nivel_Medio: True si es Medio, False si no.

Ejemplo de interpretación: Carlos
Carlos tiene:
• Nombre = Carlos
• Nivel_Educativo = Alto
• Edad = 45
• Ingreso = 3200

Transformaciones:
• Nivel_Label = 0 (porque “Alto” fue codificado como 0)
• Edad_Scale = 0.8214 (edad cercana al máximo del grupo)
• Ingreso_Scale = 1.0 (ingreso más alto del grupo)
• Edad_Std = 0.8676 → mayor que el promedio
• Ingreso_Std = 1.3968 → mucho mayor que el promedio
• Nivel_Educativo_Alto = 1.0 y Nivel_Alto = True → coinciden

In [ ]:
#  Escalamiento MinMaxScaler
scaler_minmax = MinMaxScaler()
df[['Edad_Scale', 'Ingreso_Scale']] = scaler_minmax.fit_transform(df[['Edad', 'Ingreso']])
print("\n Escalamiento Min-Max:")
print(df[['Edad', 'Edad_Scale', 'Ingreso', 'Ingreso_Scale']])


🔹 Escalamiento Min-Max:
   Edad  Edad_Scale  Ingreso  Ingreso_Scale
0    22    0.000000     1500       0.000000
1    35    0.464286     2500       0.588235
2    45    0.821429     3200       1.000000
3    28    0.214286     2100       0.352941
4    50    1.000000     2700       0.705882


In [ ]:
#  Escalamiento StandardScaler
scaler_std = StandardScaler()
df[['Edad_Std', 'Ingreso_Std']] = scaler_std.fit_transform(df[['Edad', 'Ingreso']])
print("\n🔹 Escalamiento Standard:")
print(df[['Edad', 'Edad_Std', 'Ingreso', 'Ingreso_Std']])


🔹 Escalamiento Standard:
   Edad  Edad_Std  Ingreso  Ingreso_Std
0    22 -1.349652     1500    -1.571468
1    35 -0.096404     2500     0.174608
2    45  0.867634     3200     1.396861
3    28 -0.771230     2100    -0.523823
4    50  1.349652     2700     0.523823


In [ ]:
#  Cálculo de Distancias
datos_escalados = df[['Edad_Std', 'Ingreso_Std']].values
dist_eucl = pairwise_distances(datos_escalados, metric='euclidean')
dist_manh = pairwise_distances(datos_escalados, metric='manhattan')
dist_mink = pairwise_distances(datos_escalados, metric='minkowski', p=3)

print("\n🔹 Distancia Euclidiana:")
print(dist_eucl)

print("\n🔹 Distancia Manhattan:")
print(dist_manh)

print("\n🔹 Distancia Minkowski (p=3):")
print(dist_mink)



🔹 Distancia Euclidiana:
[[0.         2.14928185 3.7050415  1.19671779 3.41708783]
 [2.14928185 0.         1.55668573 0.97118238 1.48762539]
 [3.7050415  1.55668573 0.         2.52485603 0.99726482]
 [1.19671779 0.97118238 2.52485603 0.         2.36552367]
 [3.41708783 1.48762539 0.99726482 2.36552367 0.        ]]

🔹 Distancia Manhattan:
[[0.         2.99932431 5.18561468 1.62606786 4.79459549]
 [2.99932431 0.         2.18629038 1.37325645 1.79527119]
 [5.18561468 2.18629038 0.         3.55954682 1.35505655]
 [1.62606786 1.37325645 3.55954682 0.         3.16852763]
 [4.79459549 1.79527119 1.35505655 3.16852763 0.        ]]

🔹 Distancia Minkowski (p=3):
[[0.         1.93915026 3.33386737 1.10339967 3.06759989]
 [1.93915026 0.         1.39622662 0.86535286 1.45281313]
 [3.33386737 1.39622662 0.         2.25634278 0.91949973]
 [1.10339967 0.86535286 2.25634278 0.         2.20288069]
 [3.06759989 1.45281313 0.91949973 2.20288069 0.        ]]
